src/mailroom.py

```diff
@@ -0,0 +1,122 @@
+import sys
+import traceback
+
+
+try:
+    input = raw_input
+except NameError:
+    pass
+
+
+DONORS = {
+    'kevin': [23, 34, 1000, 2849],
+    'bob': [234, 888, 9023],
+    'josh': [6, 1, 9, 5000, 9, 1, 2, 3],
+    'alice': [3, 1]
+}
+
+
+def interface():
+    '''Prompts user for input'''
+    response = input(u"Press 1 to send a thank you note\n"
+                     u"Press 2 to create a report\n"
+                     u"Type exit to leave the program")
+    while (response != '1') and (response != '2') and (response != 'exit'):
+        print(u"Please enter 1 or 2")
+        response = input(u"Press 1 to send a thank you note\n"
+                         u"Press 2 to create a report ")
+    if response == '1':
+        name = input(u"Enter a donor\'s full name or enter the "
+                     u"word 'list' to see a list of donors ")
+        name = name.lower()
+        while name == "list":
+            for key in DONORS:
+                print(key)
+            name = input(u"Enter a donor\'s full name or enter the "
+                         u"word 'list' to see a list of donors ")
+            name = name.lower()
+        donation = input(u"Enter the donation amount for {0}".format(name))
+        while not donation.isnumeric():
+            print("Please enter the donation using numbers.")
+            donation = input(u"Enter the donation amount for {0}".format(name))
+        handle_name(name, donation, DONORS)
+        handle_donation(name, donation, DONORS)
+        compose_email(name, donation)
+    elif response == '2':
+        create_report()
+    else:
+        traceback.print_exc(file=sys.stdout)
+        sys.exit(0)
+
+
+def handle_name(name, donation, DONORS):
+    if name not in DONORS:
+        DONORS[name] = []
+    return DONORS
+
+
+def handle_donation(name, donation, DONORS):
+    DONORS[name].append(int(donation))
+    return DONORS
+
+
+def compose_email(name, donation):
+    print("Dear {0}, Thank you for your generous donation of ${1}."
+          .format(name, donation))
+    interface()
+
+
+def create_report():
+    names = report_names(DONORS)
+    totals = report_totals(DONORS)
+    num_donations = report_num_donations(DONORS)
+    average = report_average(report_totals(DONORS),
+                             report_num_donations(DONORS))
+    report_list = zip(names, totals, num_donations, average)
+    report_list = sorted(report_list, key=lambda x: x[1])
+    report_list.reverse()
+    print(report_list)
+    interface()
+
+
+def report_names(DONORS):
+    list_names = []
+    for key in DONORS:
+        list_names.append(key)
+    return list_names
+
+
+def report_totals(DONORS):
+    total_donated = []
+    for i in DONORS:
+        totals = (sum(DONORS[i]))
+        total_donated.append(totals)
+    return total_donated
+
+
+def report_num_donations(DONORS):
+    num_donations = []
+    for i in DONORS:
+        num = len(DONORS[i])
+        num_donations.append(num)
+    return num_donations
```

Comment:
for all of these report generating functions, you are assuming that the names in your `DONORS` dict will come out in the same order each time.  This is not safe.  Can you think of another way you could ensure that all values are associated with the right names?